bookkeeper-server/src/main/java/org/apache/bookkeeper/client/TopologyAwareEnsemblePlacementPolicy.java

```diff
@@ -790,6 +794,113 @@ public void updateBookieInfo(Map<BookieId, BookieInfo> bookieInfoMap) {
         }
     }
 
+    @Override
+    public Map<Integer, BookieId> replaceNotAdheringPlacementPolicyBookie(List<BookieId> ensemble, int writeQuorumSize,
+            int ackQuorumSize) {
+        if (CollectionUtils.isEmpty(ensemble)) {
+            return Collections.emptyMap();
+        }
+        PlacementPolicyAdherence ensembleAdheringToPlacementPolicy = isEnsembleAdheringToPlacementPolicy(ensemble,
+                writeQuorumSize, ackQuorumSize);
+        if (PlacementPolicyAdherence.FAIL != ensembleAdheringToPlacementPolicy) {
```

Comment:
Do we need to deal with the `MEET_SOFT` type?

Reply:
Needn't do it. `MEET_SOFT` means the ensemble network location is more than `minNumZonesPerWriteQuorum`.